### PR TITLE
Add common Ranking protocol

### PR DESCRIPTION
**Pull request by shutterbug2000**
Tested on both Ultimate NES Remix and DKC:TF without issues

### CODE REVIEW

File: ranking/get_cached_top_x_ranking.go
Comment (DaniElectra) [inline ranking/get_cached_top_x_ranking.go]: Change all of these with returns please. This also reduces nesting for error checks.
Also, can you change the error returned from the first error check to `Ranking.InvalidArgument`?

```suggestion
		return nex.Errors.Ranking.InvalidArgument
```

File: ranking/protocol.go
Comment (DaniElectra) [inline ranking/protocol.go]: Reorder these so that the error is the last thing that is returned

```suggestion
	getCommonDataHandler                                     func(unique_id uint64) ([]byte, error)
```

File: ranking/get_cached_top_x_ranking.go
Comment (DaniElectra) [inline ranking/get_cached_top_x_ranking.go]: we are using Critical for database errors

```suggestion
		logger.Critical(err.Error())
```

File: ranking/get_ranking.go
Comment (DaniElectra) [inline ranking/get_ranking.go:13]: add a space at the end

```suggestion
	if commonRankingProtocol.getRankingsAndCountByCategoryAndRankingOrderParamHandler == nil {
		logger.Warning("Ranking::GetRanking missing GetRankingsAndCountByCategoryAndRankingOrderParamHandler!")
		return nex.Errors.Core.NotImplemented
	}

```

File: ranking/get_cached_top_x_rankings.go
Comment (DaniElectra) [inline ranking/get_cached_top_x_rankings.go]: Is this actually zero?

File: ranking/get_cached_top_x_rankings.go
Comment (DaniElectra) [inline ranking/get_cached_top_x_rankings.go]: We should set the expire time to some date in the far future like Dec 31st 9999, just in case

**Review by DaniElectra — CHANGES_REQUESTED** [review]

File: ranking/get_common_data.go
Comment (jonbarrow) [inline ranking/get_common_data.go]: I've noticed a couple places like this where `err` is only being checked if it *is* `nil`, and doing a success. But there is nothing being done for if it *isn't* `nil`

This should probably throw `Ranking::NotFound` if `commonDataErr != nil`. That way we follow "happy path", where the default flow is a success and these kinds of checks would be for errors and to bail early

File: ranking/get_ranking.go
Comment (jonbarrow) [inline ranking/get_ranking.go]: With Dani's suggestion to change lines 25 and 29 into returns, this check would no longer be needed so long as the check on line 28 is split into 2 checks

**Review by jonbarrow — CHANGES_REQUESTED** [review]
Dani got most of the stuff I had to say already

File: ranking/get_cached_top_x_rankings.go
Comment (shutterbug2000) [reply ranking/get_cached_top_x_rankings.go]: Left a comment in the code re: this. I'm not sure if it matters, but it's noted down either way :p

File: ranking/get_cached_top_x_rankings.go
Comment (shutterbug2000) [reply ranking/get_cached_top_x_rankings.go]: Left a comment in the code re: this. I'm not sure if it matters, but it's noted down either way :p

File: ranking/get_ranking.go
Comment (shutterbug2000) [reply ranking/get_ranking.go]: I'm a bit confused by "so long as the check on line 28 is split into 2 checks" :p

File: ranking/get_cached_top_x_ranking.go
Comment (DaniElectra) [inline ranking/get_cached_top_x_ranking.go]: This comment is very long to be on a single line

```suggestion
	// The real server sends the "CreatedTime" + 5 minutes. It doesn't change, even on subsequent requests, until after the ExpiredTime has passed (seemingly what the "cached" means). Whether we need to replicate this idk, but in case, here's a note
	pResult.ExpiredTime = nex.NewDateTime(serverTime.FromTimestamp(time.Now().UTC().Add(time.Minute * time.Duration(5))))
```

File: ranking/upload_common_data.go
Comment (DaniElectra) [inline ranking/upload_common_data.go]: The RMC response creation can be moved further down to be exactly before the `SetSuccess` function

File: ranking/protocol.go
Comment (DaniElectra) [inline ranking/protocol.go]: This is imported on some common protocols so that we can access the `DataHolder` structures. There aren't any of them for the Ranking protocol, so this can be removed

File: ranking/get_common_data.go
Comment (DaniElectra) [inline ranking/get_common_data.go]: little nitpick, the errors can all be renamed to `err` now that we are handling them with returns

```suggestion
	commonData, err := commonRankingProtocol.getCommonDataHandler(uniqueID)
	if err != nil {
		return nex.Errors.Ranking.NotFound
	}
```

File: ranking/get_common_data.go
Comment (DaniElectra) [inline ranking/get_common_data.go]: The empty space can be removed

```suggestion
	commonData, err := commonRankingProtocol.getCommonDataHandler(uniqueID)
	if err != nil {
		return nex.Errors.Ranking.NotFound
	}
```

File: ranking/get_cached_top_x_ranking.go
Comment (DaniElectra) [inline ranking/get_cached_top_x_ranking.go]: I would split the comment on multiple lines for better readability

File: ranking/get_cached_top_x_rankings.go
Comment (DaniElectra) [inline ranking/get_cached_top_x_rankings.go]: I would split the comment on multiple lines for better readability

**Review by DaniElectra — APPROVED** [review]
some minor format changes, otherwise looks good. I will wait for Jon's final review before merging

**Comment by jonbarrow** [comment]
LGTM :+1:

**Review by jonbarrow — APPROVED** [review]

**Pull request merged**
